homeassistant/components/intent/timers.py

```diff
@@ -0,0 +1,772 @@
+        )
+
+    async def _timer_finished(self, timer_id: str) -> None:
+        """Call event handlers when a timer finishes."""
+        timer = self.timers.pop(timer_id, None)
+        if timer is None:
+            return
+
+        # Force seconds left to 0
+        timer.seconds = 0
+        timer.updated_at = time.monotonic_ns()
+        self.hass.bus.async_fire(EVENT_INTENT_TIMER_FINISHED, timer.to_event())
+
+        _LOGGER.debug(
+            "Timer finished: id=%s, name=%s, device_id=%s",
+            timer_id,
+            timer.name,
+            timer.device_id,
+        )
+
+
+def _find_timer(hass: HomeAssistant, slots: dict[str, Any]) -> TimerInfo:
+    """Match a single timer with constraints or raise an error."""
+    timer_manager: TimerManager = hass.data[TIMER_DATA]
+    matching_timers: list[TimerInfo] = list(timer_manager.timers.values())
+    has_filter = False
+
+    # Search by name first
+    name: str | None = None
+    if "name" in slots:
+        has_filter = True
+        name = slots["name"]["value"]
+        assert name is not None
+        name_norm = name.strip().casefold()
+
+        matching_timers = [t for t in matching_timers if t.name_normalized == name_norm]
+        if len(matching_timers) == 1:
+            # Only 1 match
+            return matching_timers[0]
+
+    # Use starting time to disambiguate
+    start_hours: int | None = None
+    if "start_hours" in slots:
+        start_hours = int(slots["start_hours"]["value"])
+
+    start_minutes: int | None = None
+    if "start_minutes" in slots:
+        start_minutes = int(slots["start_minutes"]["value"])
+
+    start_seconds: int | None = None
+    if "start_seconds" in slots:
+        start_seconds = int(slots["start_seconds"]["value"])
+
+    if (
+        (start_hours is not None)
+        or (start_minutes is not None)
+        or (start_seconds is not None)
+    ):
+        has_filter = True
+        matching_timers = [
+            t
+            for t in matching_timers
+            if (t.start_hours == start_hours)
+            and (t.start_minutes == start_minutes)
+            and (t.start_seconds == start_seconds)
+        ]
+
+        if len(matching_timers) == 1:
+            # Only 1 match remaining
+            return matching_timers[0]
+
+    if (not has_filter) and (len(matching_timers) == 1):
+        # Only 1 match remaining with no filter
+        return matching_timers[0]
+
+    # Use device id
+    device_id: str | None = None
+    if matching_timers and ("device_id" in slots):
+        device_id = slots["device_id"]["value"]
+        assert device_id is not None
+        matching_device_timers = [
+            t for t in matching_timers if (t.device_id == device_id)
+        ]
+        if len(matching_device_timers) == 1:
+            # Only 1 match remaining
+            return matching_device_timers[0]
+
+        # Try area/floor
+        if (
+            (device := timer_manager.device_registry.async_get(device_id))
```

Comment:
No need to keep a handle to device_registry just for this invocation. You can fetch it here.